README.md
# spotilocal
Unofficial api for Spotify's local web server.

## How it works
It uses the same method as https://developer.spotify.com/technologies/widgets/spotify-play-button/ does. 
Basically Spotify desktop runs a web server on 127.0.0.1 that can accept specific commands and even open 
Spotify client itself.  

Spotify domain spotilocal.com that simply points to 127.0.0.1. Apparently in order to bypass browser's 
limitation of amount of concurrent connection to the same domain they use randomly generated subdomains 
of spotilocal.com. 

## Api

### Spotilocal#init(defaultPort?:number)

Initializes the Spotilocal client. Returns a Promise for Spotilocal client.

defaultPort - Port to try first.

### Spotilocal#getStatus(returnOn?:Array<`play`|`pause`|`login`|`logout`|`error`|`ap`>, returnAfter?:number)

Returns a Promise for Status.

returnOn - Activates long-polling to return when one or more specific event(s) occur.

returnAfter - Defines a timeout. Pass 0 to disable. Default is 0.

### Spotilocal#pause(pause:boolean)

Pauses or unpauses current playback. Returns a Promise for Status.  

### Spotilocal#play(uri:string, context?:string)

Plays song with uri in provided context.

uri - track uri.

context - of song(where it exists). Examples: playlist uri, album uri and so on. 

Returns a Promise for Status. 

## Example usage 


```ts
import {Spotilocal} from 'spotilocal';

const spotilocal = new Spotilocal();
const trackUri = 'spotify:track:3cANM3NuUjRDTi8fdU8P6q';
spotilocal.init().then((spotilocal) => {
    return spotilocal.play(trackUri, 'spotify:user:shyyko.serhiy:playlist:4SdN0Re3tJg9uG08z2Gkr1')
}).then((status) => {
    console.log(`Playing:   ${status.playing}`);
    console.log(`Song Name: ${status.track.track_resource.name}`);
    return spotilocal.pause(true);
}).catch(console.error);
```




CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.4.4"></a>
## [0.4.4](https://github.com/ShyykoSerhiy/spotilocal/compare/v0.4.3...v0.4.4) (2017-10-10)


### Bug Fixes

* changelog ([d82a688](https://github.com/ShyykoSerhiy/spotilocal/commit/d82a688))



<a name="0.4.3"></a>
## [0.4.3](https://github.com/ShyykoSerhiy/spotilocal/compare/v0.4.2...v0.4.3) (2017-10-10)


### Bug Fixes

* added getSpotilocalUrlByPort to fix code duplication ([7c0b950](https://github.com/ShyykoSerhiy/spotilocal/commit/7c0b950))

### Features

* Added defaultPort param to init method


<a name="0.4.2"></a>
## [0.4.2](https://github.com/ShyykoSerhiy/spotilocal/compare/v0.4.1...v0.4.2) (2017-09-25)



<a name="0.4.1"></a>
## [0.4.1](https://github.com/ShyykoSerhiy/spotilocal/compare/v0.4.0...v0.4.1) (2017-09-25)


### Bug Fixes

* spotify has changed ports from 4370-4380 to 4380-4389 ([f23bb15](https://github.com/ShyykoSerhiy/spotilocal/commit/f23bb15))



<a name="0.4.0"></a>
# 0.4.0 (2017-09-24)


### Bug Fixes

* changed getStatus returnOn param type to array of ReturnOnParam ([1c69cf3](https://github.com/ShyykoSerhiy/spotilocal/commit/1c69cf3))
* changed song in tests(Stuck in the Sound - Brother is not in the spotify library for some reason) ([02e5426](https://github.com/ShyykoSerhiy/spotilocal/commit/02e5426))


### Features

* added SPOTILOCAL_IS_NOT_RUNNING string ([a02e8a4](https://github.com/ShyykoSerhiy/spotilocal/commit/a02e8a4))



<a name="0.3.0"></a>
# 0.3.0 (2017-09-24)


### Bug Fixes

* changed getStatus returnOn param type to array of ReturnOnParam ([1c69cf3](https://github.com/ShyykoSerhiy/spotilocal/commit/1c69cf3))
* changed song in tests(Stuck in the Sound - Brother is not in the spotify library for some reason) ([02e5426](https://github.com/ShyykoSerhiy/spotilocal/commit/02e5426))


### Features

* added SPOTILOCAL_IS_NOT_RUNNING string ([a02e8a4](https://github.com/ShyykoSerhiy/spotilocal/commit/a02e8a4))


test/test.ts
import { Spotilocal, SPOTILOCAL_IS_NOT_INITIALIZED, ReturnOnParam } from '../src/index';
import { assert } from 'chai';

interface IWithTimeout {
    timeout: (timeout: number) => void
}

describe('#init()', function (this: IWithTimeout) {
    this.timeout(10000);
    it('should init spotilocal', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.init().then(() => { done(); }).catch(done);
    });
});

describe('#getStatus()', function (this: IWithTimeout) {
    this.timeout(10000);
    it('should fail if not initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.getStatus().then(() => {
            done('Should have failed');
        }).catch((error) => {
            assert.strictEqual(error, SPOTILOCAL_IS_NOT_INITIALIZED);
            done();
        }).catch((error) => {
            done(error);
        });
    });

    it('should get status from spotilocal if initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.init().then((spotilocal) => {
            return spotilocal.getStatus()
        }).then((_status) => {
            done();
        }).catch(done);
    });
});

describe('#pause()', function (this: IWithTimeout) {
    this.timeout(10000);
    it('should fail if not initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.pause(true).then(() => {
            done('Should have failed');
        }).catch((error) => {
            assert.strictEqual(error, SPOTILOCAL_IS_NOT_INITIALIZED);
            done();
        }).catch((error) => {
            done(error);
        });
    });

    it('should get status from spotilocal if initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.init().then((spotilocal) => {
            return spotilocal.pause(true)
        }).then((_status) => {
            done();
        }).catch(done);
    });
});

describe('#play()', function (this: IWithTimeout) {
    this.timeout(10000);
    it('should fail if not initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.play('spotify:track:23r4eXV6ziw0NNznZU9NiC', 'spotify:user:shyyko.serhiy:playlist:4SdN0Re3tJg9uG08z2Gkr1').then(() => {
            done('Should have failed');
        }).catch((error) => {
            assert.strictEqual(error, SPOTILOCAL_IS_NOT_INITIALIZED);
            done();
        }).catch((error) => {
            done(error);
        });
    });

    it('should play song wihtout context and return correct status', function (done) {
        const spotilocal = new Spotilocal();
        const trackUri = 'spotify:track:53zHrgxt8Xy1RkMWepJVUh';
        spotilocal.init().then((spotilocal) => {
            return spotilocal.play(trackUri)
        }).then((status) => {
            assert.strictEqual(status.playing, true);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            assert.strictEqual(status.track.track_resource.name, 'Insane');
            done();
        }).catch(done);
    });

    it('should play song in context and return correct status', function (done) {
        const spotilocal = new Spotilocal();
        const trackUri = 'spotify:track:7H2zem6ynyk67mqqXDIwDS';
        spotilocal.init().then((spotilocal) => {
            return spotilocal.play(trackUri, 'spotify:user:shyyko.serhiy:playlist:4SdN0Re3tJg9uG08z2Gkr1')
        }).then((status) => {
            assert.strictEqual(status.playing, true);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            assert.strictEqual(status.track.track_resource.name, 'Creature Fear');
            done();
        }).catch(done);
    });
});

describe('#play() and #pause()', function (this: IWithTimeout) {
    this.timeout(10000);
    it('should play song pause it and resume it', function (done) {
        const spotilocal = new Spotilocal();
        const trackUri = 'spotify:track:23r4eXV6ziw0NNznZU9NiC';
        spotilocal.init().then((spotilocal) => {
            return spotilocal.play(trackUri, 'spotify:user:shyyko.serhiy:playlist:4SdN0Re3tJg9uG08z2Gkr1')
        }).then((status) => {
            assert.strictEqual(status.playing, true);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            return spotilocal.pause(true);
        }).then((status) => {
            assert.strictEqual(status.playing, false);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            return spotilocal.pause(false);
        }).then((status) => {
            assert.strictEqual(status.playing, true);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            return spotilocal.pause(true);
        }).then((status) => {
            assert.strictEqual(status.playing, false);
            assert.strictEqual(status.track.track_resource.uri, trackUri);
            done();
        }).catch(done);
    });
});

describe('#getStatus() with returnOn', function (this: IWithTimeout) {
    this.timeout(10000);
    const returnOn: ReturnOnParam[] = ['play', 'pause'];
    it('should fail if not initialized', function (done) {
        const spotilocal = new Spotilocal();
        spotilocal.getStatus(returnOn).then(() => {
            done('Should have failed');
        }).catch((error) => {
            assert.strictEqual(error, SPOTILOCAL_IS_NOT_INITIALIZED);
            done();
        }).catch((error) => {
            done(error);
        });
    });

    it('should get status from spotilocal if initialized', function (done) {
        const spotilocal = new Spotilocal();
        let start = 0;
        spotilocal.init().then((spotilocal) => {
            start = +new Date();
            setTimeout(() => {
                spotilocal.play('spotify:track:23r4eXV6ziw0NNznZU9NiC', 'spotify:user:shyyko.serhiy:playlist:4SdN0Re3tJg9uG08z2Gkr1')
            }, 1000)
            return spotilocal.getStatus(returnOn)
        }).then((status) => {
            assert.strictEqual(status.playing, true);            
            assert.isAbove(+new Date() - start, 1000);            
            done();
        }).catch(done);
    });
});


src/status.ts
export interface Status {
    "version": number, // 9 
    "client_version": string, //"1.0.36.124.g1cba1920", 
    "playing": boolean,
    "shuffle": boolean,
    "repeat": boolean,
    "play_enabled": boolean,
    "prev_enabled": boolean,
    "next_enabled": boolean,
    "track": {
        "track_resource": {
            "name": string,//"Ready", 
            "uri": string,//"spotify:track:5FPctWaZw32JIRJrOtW8fp", 
            "location": {
                "og": string//"https://open.spotify.com/track/5FPctWaZw32JIRJrOtW8fp"
            }
        },
        "artist_resource": {
            "name": string,//"Sondar", 
            "uri": string,//"spotify:artist:18QL43RyFNcvKqvL1uYtpz", 
            "location": {
                "og": string//"https://open.spotify.com/artist/18QL43RyFNcvKqvL1uYtpz"
            }
        },
        "album_resource": {
            "name": string,//"Ready", 
            "uri": string,//"spotify:album:4wX476IXAup6Oo7PBN6AvE", 
            "location": {
                "og": string//"https://open.spotify.com/album/4wX476IXAup6Oo7PBN6AvE"
            }
        },
        /**
         * Duration of the track
         */
        "length": number,//207,   
        "track_type": string//"normal"
    },
    "context": {},
    "playing_position": number,//63.02, 
    "server_time": number,//1473180693, 
    "volume": number,//0.88999772, 
    "online": boolean,//true, 
    "open_graph_state": {
        "private_session": boolean,//false, 
        "posting_disabled": boolean,//false
    },
    "running": boolean//true
}

src/index.ts
import * as request from 'superagent';
import {Status} from './status';

function failIfNotInitialized(_target: any, _propertyKey: string, descriptor: PropertyDescriptor):PropertyDescriptor{
    const fn = descriptor.value as Function;

    if (typeof fn !== 'function'){
        throw new Error(`@failIfNotInitialized can only be applied to method and not to ${typeof fn}`)
    }

    return Object.assign({}, descriptor, {value: function(this: Spotilocal){
        if (!this.initialized) {
            return Promise.reject<string>(SPOTILOCAL_IS_NOT_INITIALIZED);
        }         
        return fn.apply(this, arguments);
    }})

} 

export const SPOTILOCAL_IS_NOT_INITIALIZED = 'Spotilocal is not initialized';
export const SPOTILOCAL_IS_NOT_RUNNING = 'It looks like Spotify isn\'t open. We failed to find spotilocal url with ports in range 4370-4380.';
export const RETURN_ON_PLAY = 'play' as 'play';
export const RETURN_ON_PAUSE = 'pause' as 'pause';
export const RETURN_ON_LOGIN = 'login' as 'login';
export const RETURN_ON_LOGOUT = 'logout' as 'logout';
export const RETURN_ON_ERROR = 'error' as 'error';
export const RETURN_ON_AP = 'ap' as 'ap';

export type ReturnOnParam = typeof RETURN_ON_PLAY | typeof RETURN_ON_PAUSE | typeof RETURN_ON_LOGIN | typeof RETURN_ON_LOGOUT | typeof RETURN_ON_ERROR | typeof RETURN_ON_AP;

export class Spotilocal {
    public initialized: boolean;
    private spotilocalUrl: string;
    private _port: number;
    public get port(): number {
        return this._port;
    }
    private oauth: string;
    private csrf: string;    

    constructor() {
    }

    public init(defaultPort?: number): Promise<Spotilocal> {
        return Spotilocal.getSpotilocalUrl(defaultPort).then(port => {
            this.spotilocalUrl = Spotilocal.getSpotilocalUrlByPort(port);
            this._port = port;
            return Promise.all([Spotilocal.getOauthToken(), Spotilocal.getCsrfToken(this.spotilocalUrl)]);
        }).then((value) => {
            this.oauth = value[0];
            this.csrf = value[1];
            this.initialized = true;
            return this;
        })
    }

    @failIfNotInitialized
    public getStatus(returnOn?: ReturnOnParam[], returnAfter: number = 0): Promise<Status> {
        let timeout = 1000;

        const params = new Map<string, any>();
        if (returnOn && returnOn.length) {
            params.set('returnon', returnOn.join(','));
            params.set('returnafter', returnAfter); // 0 disables the timeout, passing -1 can cause spontaneous high CPU usage

            if (returnAfter === 0) timeout = 0;
            else timeout = (returnAfter * 1000) + 1000;
        }

        return this.genericCommand('status', params, timeout);
    }

    /**
     * Pauses(or unpauses) playback of spotify.
     * @param pause if true, then pauses playback. If else resumes playback.
     */
    @failIfNotInitialized
    public pause(pause:boolean = true): Promise<Status> {        
        const params = new Map<string, any>();
        params.set('pause', pause);     
        return this.genericCommand('pause', params);           
    }

    /**
     * Plays song with uri in provided context.
     * @param uri track uri
     * @param context context of song(where it exists). Examples: playlist uri, album uri and so on. 
     */
    @failIfNotInitialized    
    public play(uri:string, context?:string): Promise<Status> {        
        const params = new Map<string, any>();
        params.set('uri', uri);
        context && params.set('context', context);     
        return this.genericCommand('play', params);           
    }  

    private genericCommand(command: string, additionalProps?: Map<string, any>, timeout?: number): Promise<Status> {
        const additionalQuery = (additionalProps && additionalProps.size) ? `&${Array.from(additionalProps.entries()).reduce((prev, curr) => {
            return `${prev}&${curr[0]}=${encodeURIComponent(curr[1])}`
        }, '')}` : '';
        return new Promise((resolve, reject) => {
            Spotilocal.requestToAbsolutelyUglyNotSecuredRequest(
                request.get(`${this.spotilocalUrl}remote/${command}.json?csrf=${this.csrf}&oauth=${this.oauth}${additionalQuery}`), timeout)
                .end((err, res) => {
                    if (err || !res.ok) {
                        reject(err || res.status);
                    } else {
                        resolve(res.body as Status);
                    }
                })
        });
    }

    /**
     * Gets oauth token that will be used in all later calls to spotilocal
     */
    public static getOauthToken(): Promise<string> {
        return new Promise((resolve, reject) => {
            request.get('https://open.spotify.com/token').end((err, res) => {
                if (err || !res.ok) {
                    reject(err || res.status);
                } else {
                    var t = (res.body as { t: string }).t;
                    resolve(t);
                }
            })
        });
    }

    /**
     * Gets csrf token that will be used in all later calls to spotilocal
     */
    public static getCsrfToken(spotilocalUrl: string): Promise<string> {
        return new Promise((resolve, reject) => {
            Spotilocal.requestToAbsolutelyUglyNotSecuredRequest(request.get(`${spotilocalUrl}simplecsrf/token.json`))
                .end((err, res) => {
                    if (err || !res.ok) {
                        reject(err || res.status);
                    } else {
                        var token = (res.body as { token: string }).token;
                        resolve(token);
                    }
                })
        });
    }

    /**
     * Gets spotilocal api url with port in range 4370-4389.
     */
    public static getSpotilocalUrl(defaultPort?: number): Promise<number> {
        return new Promise((resolve, reject) => {
            const tryGetSpotilocalVersion = (port: number) => {
                if (port > 4389) {
                    reject(SPOTILOCAL_IS_NOT_RUNNING);
                    return;
                }
                Spotilocal.getSpotilocalVersion(Spotilocal.getSpotilocalUrlByPort(port)).then(() => { resolve(port) }).catch(() => {
                    tryGetSpotilocalVersion(port + 1);
                });
            }

            if (defaultPort) {
                Spotilocal.getSpotilocalVersion(Spotilocal.getSpotilocalUrlByPort(defaultPort)).then(() => { resolve(defaultPort) }).catch(() => {
                    tryGetSpotilocalVersion(4370);
                });
            } else {
                tryGetSpotilocalVersion(4370);
            }
        });
    }

    /**
     * Gets spotilocal version (presumably version of spotify local server)
     */
    public static getSpotilocalVersion(apiUrl: string): Promise<string> {
        return new Promise((resolve, reject) => {
            Spotilocal.requestToAbsolutelyUglyNotSecuredRequest(request.get(`${apiUrl}service/version.json?service=remote`))
                .end((err, res) => {
                    if (err || !res.ok) {
                        reject(err || res.status);
                    } else {
                        resolve(JSON.parse(res.text));
                    }
                });
        });
    }

    /**
     * Sets rejectUnauthorized to false, Origin to https://open.spotify.com and timeout to 1000
     */
    public static requestToAbsolutelyUglyNotSecuredRequest(request: request.SuperAgentRequest, timeout: number = 1000): request.SuperAgentRequest {
        const req = request.set('Origin', 'https://open.spotify.com');
        if (timeout !== 0) { // Disable timeout if set to 0
            req.timeout(timeout);
        }

        return req;
    }

    private static getSpotilocalUrlByPort(port: number) {
        return `http://127.0.0.1:${port}/`;
    }
}
